### PR TITLE
Final changes for mfem-4.1

### PR DESCRIPTION
**Pull request by tzanio**
### Release Checklist

- [x] Update the MFEM version in the following files:
    - [x] `CHANGELOG`
    - [x] `makefile`
    - [x] `CMakeLists.txt`
    - [x] `doc/CodeDocumentation.conf.in`
- [x] (LLNL only) Make sure all `README.html` files in the source repo are up to date.
- [ ] Tag the repository:

  ```
  git tag -a v4.1 -m "Official release v4.1"
  git push origin v4.1
  ```
- [ ] Create the release tarball and push to `mfem/releases`.
- [ ] Recreate the `next` branch as described in previous section.
- [ ] Update and push documentation  to `mfem/doxygen`.
- [ ] Update URL shorlinks:
    - [ ] Create a shortlink at [https://goo.gl/](https://goo.gl/) for the release tarball, e.g. http://mfem.github.io/releases/mfem-4.1.tgz.
    - [ ] (LLNL only) Add and commit the new shorlink in the `links` and `links-mfem` files of the internal `mfem/downloads` repo.
    - [ ] Add the new shortlinks to the MFEM packages in `spack`, `homebrew/science`, `VisIt`, etc.
- [ ] Update website in `mfem/web` repo:
    - Update version and shortlinks in `src/index.md` and `src/download.md`.
    -  Use [cloc-1.62.pl](http://cloc.sourceforge.net/) and `ls -lh` to estimate the SLOC and the tarball size in `src/download.md`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1314](https://github.com/mfem/mfem/pull/1314) | @tzanio | @tzanio | @v-dobrev + @camierjs + @YohannDudouit | 02/22/20 | 03/05/20 | 03/10/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Besides ef10ba8074e03dcded82ad4f4dbc38f7c68d4d59, should we mention/document the non-determinisim anywhere else?

**Comment by camierjs** [comment]
Perhaps in the CEED kernels headers, the ones inside MFEM?

**Comment by tzanio** [comment]
> Perhaps in the CEED kernels headers, the ones inside MFEM?

@YohannDudouit, can you make this change?

**Comment by YohannDudouit** [comment]
I added a note in `general/device.cpp` where the default libCEED GPU backend is selected.

File: general/device.cpp [hunk 1]
```diff
@@ -330,6 +330,8 @@ void Device::Setup(const int device)
    {
       if (!ceed_option)
       {
+         // NOTE: The current default libCEED GPU backend (ceed-cuda) uses
+         // atomics and therefore is non-deterministic!
```
Comment (tzanio) [inline general/device.cpp]:
```suggestion
         // NOTE: libCEED's /gpu/cuda/gen backend (ceed-cuda) is non-deterministic!
```

File: general/device.hpp [hunk 2]
```diff
@@ -60,7 +60,9 @@ struct Backend
       CEED_CPU  = 1 << 10,
       /** @brief [device] CEED CUDA backend working in colaboration with the
           CUDA backend. Enabled when MFEM_USE_CEED = YES and
-          MFEM_USE_CUDA = YES. */
+          MFEM_USE_CUDA = YES.
+          NOTE: The current default libCEED GPU backend (ceed-cuda) uses
+          atomics and therefore is non-deterministic! */
```
Comment (tzanio) [inline general/device.hpp]:
```suggestion
          NOTE: The current default libCEED GPU backend is non-deterministic! */
```

**Comment by IdoAkkerman** [comment]
@tzanio the checklists says v3.1 instead of v4.1. Perhaps better to modify, to prevent mistakes?

**Comment by tzanio** [comment]
@YohannDudouit, can you add the sample run for the simplex ceed backend that we discussed in this branch?

**Comment by tzanio** [comment]
> @YohannDudouit, can you add the sample run for the simplex ceed backend that we discussed in this branch?

@YohannDudouit, reminder

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: tests/unit/fem/test_pa_coeff.cpp [hunk 3]
```diff
@@ -57,7 +57,7 @@ TEST_CASE("H1 pa_coeff")
             std::cout << "Testing " << dimension << "D partial assembly with "
                       << "coeffType " << coeffType << " and "
                       << "integrator " << integrator << std::endl;
-            for (int order = 1; order < 5; ++order)
+            for (int order = 1; order < 4; ++order)
             {
```
Comment (tzanio) [inline tests/unit/fem/test_pa_coeff.cpp]: @dylan-copeland, any objection to reduce the order in this test?

It speeds-up the unit testing significantly. Compare for example these Travis results ([old](https://travis-ci.org/mfem/mfem/builds/659637428) vs [new](https://travis-ci.org/mfem/mfem/builds/659691060)):

- **Linux Serial:** [old, 19.24s](https://travis-ci.org/mfem/mfem/jobs/659637433#L694) vs [new, 6.05s](https://travis-ci.org/mfem/mfem/jobs/659691065#L695)
- **Linux Parallel:** [old, 20.81s](https://travis-ci.org/mfem/mfem/jobs/659637435#L989) vs [new, 7.00s](https://travis-ci.org/mfem/mfem/jobs/659691067#L991)
- **Mac Serial:** [old, 25.02s](https://travis-ci.org/mfem/mfem/jobs/659637437#L523) vs [new, 8.27s](https://travis-ci.org/mfem/mfem/jobs/659691069#L524)
- **Mac Parallel:** [old, 66.77s](https://travis-ci.org/mfem/mfem/jobs/659637439#L664) vs [new 18.81s](https://travis-ci.org/mfem/mfem/jobs/659691071#L662)

Comment (tzanio) [reply tests/unit/fem/test_pa_coeff.cpp]: @camierjs, in the above runs, do we expect the time for `mpirun -np 4 sedov_tests_debug` in parallel and `sedov_tests_debug` in serial to be relatively high?

Comment (camierjs) [reply tests/unit/fem/test_pa_coeff.cpp]: Yes, I have looked at this issue in another branch to speed it up, especially on MacOS.

Comment (dylan-copeland) [reply tests/unit/fem/test_pa_coeff.cpp]: @tzanio No objections. If something breaks the lower orders should hopefully catch it.

**Comment by tzanio** [comment]
@mfem/developers please verify the mfem-4.1 updates in the `CHANGELOG` for correctness and completeness.

**Comment by v-dobrev** [comment]
The 64-bit build seems to be OK now -- I don't see issues running `make test`.

However, there seems to be an issue with `ex19p` and `ex21p` on quartz with the latest hypre (v2.18.2); with older version of hypre (v2.11.2), the same examples work fine.

File: CHANGELOG [hunk 4]
```diff
@@ -59,9 +59,35 @@ libCEED support
...
   libCEED, see the sample runs in examples 1 and 6.
 
+- NOTE: The current default libCEED GPU backend (ceed-cuda) uses atomics and
+  therefore is non-deterministic.
+
+Partial assembly and matrix-free discretizations
+------------------------------------------------
+- The support for matrix-free methods on both CPU and GPU devices based on a
+  partially assembled operator decomposition was extended to include:
+  * DG integrarors, (for now only in the Gauss-Lobatto basis), see Example 9,
```
Comment (dylan-copeland) [inline CHANGELOG]: DG integrators

File: CHANGELOG [hunk 5]
```diff
@@ -59,9 +59,35 @@ libCEED support
...
+  * vector mass and vector diffusion bilinear integrators,
+  * convection integrator with improved performance,
+  * gradient and vector divergence integrators for Stokes problems,
+  * initial partial assembly mode for NonlinearForms.
+
+- Diagonals of partially assembled operators can now be computed efficiently.
+  See the new methods AssembleDiagonal in BilinearForm, AssembleDiagonalPA in
+  BilinearFormIntegrator and the implementations in fem/bilininteg_*.cpp.
+
+- In many examples, the partial assembly algorithms provides significantly
```
Comment (dylan-copeland) [inline CHANGELOG]: algorithms provide

File: CHANGELOG [hunk 6]
```diff
@@ -127,76 +138,76 @@ Discretization improvements
...
-- Added support for HiOp, a lightweight HPC solver for nonlinear optimization
-  problems see class HiOpNLPOptimizer and the example codes in examples/hiop.
-
 - Added support for Ginkgo, a high-performance linear algebra library for GPU
   and manycore nodes, with a focus on sparse solution of linear systems. For
   more details see linalg/ginkgo.hpp and the example code in examples/gingko.
 
-- Added Adams-Bashforth and Adams-Moulton time integrators.
+- Added support for HiOp, a lightweight HPC solver for nonlinear optimization
+  problems see class HiOpNLPOptimizer and the example codes in examples/hiop.
```
Comment (dylan-copeland) [inline CHANGELOG]: problems; see

**Comment by tzanio** [comment]
Thanks for the CHANGELOG fixes Dylan.

Anyone else had a chance to look at it? Direct link: https://github.com/mfem/mfem/blob/mfem-4.1-dev/CHANGELOG

**Pull request merged**